pkg/controller/storagecluster/reconcile.go

```diff
@@ -121,6 +123,33 @@ func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcil
 		}
 	}
 
+	// Check GetDeletionTimestamp to determine if the object is under deletion
+	if instance.GetDeletionTimestamp().IsZero() {
+		// The object is not being deleted, so let's add our finalizer
+		if !contains(instance.GetFinalizers(), storageClusterFinalizer) {
+			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, storageClusterFinalizer)
+			if err := r.client.Update(context.TODO(), instance); err != nil {
+				reqLogger.Error(err, "Failed to update storagecluster with finalizer")
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is marked for deletion
+		if contains(instance.GetFinalizers(), storageClusterFinalizer) {
+			if err := r.deleteResources(instance); err != nil {
+				// If the dependencies failed to delete, retry again
+				return reconcile.Result{}, err
+			}
+			// Once all finalizers have been removed, the object will be deleted
+			instance.ObjectMeta.Finalizers = remove(instance.ObjectMeta.Finalizers, storageClusterFinalizer)
+			if err := r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		reqLogger.Info("Object is terminated, skipping reconciliation")
+		return reconcile.Result{}, nil
+	}
+
```

Comment:
this doesn't matter much, but you could just always ensure the finalizer exists here after the check for deletion since the deletion check returns early.

Reply:
Not sure I quite understand you there. Are you saying check for deletion and then ensure the presense/absense of finalizer?